examples/fluids/src/dirichlet.c

```diff
@@ -0,0 +1,194 @@
+
+    CeedOperatorSetField(op_dirichlet_sub, "surface qdata", elem_restr_qd_sur,
+                         CEED_BASIS_COLLOCATED, q_data_sur);
+    CeedOperatorSetField(op_dirichlet_sub, "x", elem_restr_x_stored,
+                         CEED_BASIS_COLLOCATED, x_stored);
+    CeedOperatorSetField(op_dirichlet_sub, "scale", elem_restr_scale,
+                         CEED_BASIS_COLLOCATED, scale_stored);
+    CeedOperatorSetField(op_dirichlet_sub, "q", elem_restr_q_sur,
+                         CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+    CeedOperatorSetNumQuadraturePoints(op_dirichlet_sub, elem_size);
+
+    // -- Add to composite operator
+    CeedCompositeOperatorAddSub(op_dirichlet, op_dirichlet_sub);
+
+    CeedVectorDestroy(&q_data_sur);
+    CeedVectorDestroy(&multiplicity);
+    CeedVectorDestroy(&x_stored);
+    CeedVectorDestroy(&scale_stored);
+    CeedElemRestrictionDestroy(&elem_restr_x_sur);
+    CeedElemRestrictionDestroy(&elem_restr_q_sur);
+    CeedElemRestrictionDestroy(&elem_restr_qd_sur);
+    CeedElemRestrictionDestroy(&elem_restr_x_stored);
+    CeedElemRestrictionDestroy(&elem_restr_scale);
+    CeedQFunctionDestroy(&qf_strongbc);
+    CeedOperatorDestroy(&op_setup_sur);
+    CeedOperatorDestroy(&op_dirichlet_sub);
+    CeedOperatorDestroy(&op_setup);
+  }
+
+  CeedBasisDestroy(&basis_x_to_q_sur);
+  CeedQFunctionDestroy(&qf_setup);
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode DMPlexInsertBoundaryValues_StrongBCCeed(DM dm,
+    PetscBool insert_essential, Vec Q_loc, PetscReal time, Vec face_geom_FVM,
+    Vec cell_geom_FVM, Vec grad_FVM) {
+  Vec boundary_mask;
+  User user;
+  PetscScalar *q;
+  PetscMemType q_mem_type;
+  PetscFunctionBeginUser;
+
+  PetscCall(DMGetApplicationContext(dm, &user));
+
+  // Mask Dirichlet entries
+  PetscCall(DMGetNamedLocalVector(dm, "boundary mask", &boundary_mask));
+  PetscCall(VecPointwiseMult(Q_loc, Q_loc, boundary_mask));
+  PetscCall(DMRestoreNamedLocalVector(dm, "boundary mask", &boundary_mask));
+
+  // Setup libCEED vector
+  PetscCall(VecGetArrayAndMemType(Q_loc, &q, &q_mem_type));
+  CeedVectorSetArray(user->q_ceed, MemTypeP2C(q_mem_type), CEED_USE_POINTER, q);
+
+  // Apply libCEED operator
+  CeedOperatorApplyAdd(user->op_dirichlet, CEED_VECTOR_NONE, user->q_ceed,
+                       CEED_REQUEST_IMMEDIATE);
+
+  // Restore PETSc vectors
+  CeedVectorTakeArray(user->q_ceed, MemTypeP2C(q_mem_type), NULL);
+  PetscCall(VecRestoreArrayAndMemType(Q_loc, &q));
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode SetupStrongBC_Ceed(Ceed ceed, CeedData ceed_data, DM dm,
+                                  User user, AppCtx app_ctx, ProblemData *problem,
+                                  SimpleBC bc, CeedInt Q_sur, CeedInt q_data_size_sur) {
+  PetscFunctionBeginUser;
+
+  {
+    Vec boundary_mask, global_vec;
+
+    PetscCall(DMGetNamedLocalVector(dm, "boundary mask", &boundary_mask));
+    PetscCall(DMCreateGlobalVector(dm, &global_vec));
```

Comment:
I've currently set for a global vector to be created and destroyed to get the boundary mask value. I could alternatively use the `Q` vector created in the `main` function, but that'd involve needing to pass it through several functions or adding it to a struct. This is currently run before `Q` is created, so there aren't two copies of a global vector. 

Wasn't sure if having two "utility" global vectors would actually take up space; Not entirely sure how global vs local vectors work on the backend (ie. whether PETSc knows that the only value in `global_vec` is just 1 and doesn't need to allocate a full array for that).

Reply:
PETSc doesn't optimize that case, but you're doing this once so it's totally reasonable. You can use `DMGetGlobalVector` to have it be a work vector that doesn't get deallocated when returned (so the the next caller can have it). I think either is fine here.